### ISSUE TITLE
Explore a wrapper/plugin/feature to handle JSON API 1.0

### ISSUE THREAD

**Issue by jmdobry** [js-data/js-data#169]

**Comment by DevBlaaast** [comment]
Hey @jmdobry ! I've been doing some work on Ember Data and I use your plugin for my Ionic apps :) thanks for it !

I have a nice experience so far (though I'm not doing complicated stuffs), I find a lot of concepts from Ember Data which is good. I could be interested in working with you on a JSON API 1.0 store/serializer/adapter. Let me know how you want to handle that (and if you're interested of course) !

Cheers !

**Comment by BlairAllegroTech** [comment]
Yes we are wanting to add support for JSON API also and would be interesting in participating

**Comment by jmdobry** [comment]
K, this is what I think needs to happen here. A new project needs to be created under the js-data org, say `js-data-jsonapi`, which will be a superset of the datastore. Basically, it will wrap js-data with some extra functionality. With my limited knowledge of the JSON API spec, the wrapper probably just needs a way to define resources automatically based on what it finds as it's parsing the API, store more metadata as it parsing the API, and then exposing methods so you can navigate your way around the API.

Does someone with more knowledge of the JSON API want to add to this?

**Comment by BlairAllegroTech** [comment]
Yes, as JSON API follows HATEOAS quite strictly it should be possible to discover the API schema dynamically as the system traverses these relationships.This would be a nice approach. 

What I was about to start experimenting with was more basic than that. I was intending to use the existing js-data schema definition but to try to add the ability to lazy load resources. This is because JSON API defines relationship links which are discovered as the API is traversed, but does not necessarily provide the data for those related resources (Optionally provided in included data section). 
So after requesting a resource, we would also need to add uninitialised related objects to the data store along with the Url that they should be initialised from. These could be considered lazy objects which would need to be initialized from their resource Url from the backing store / adapter.

The Url and status of the object e.g. Loaded or not for lazy objects could possibly be stored in metadata? I am new to js-data so am not sure if this is possible ?

An modified Http Adapter could then inspect this metadata and load the lazy objects as required ?

Once we have the data and their resource Urls stored we should have all the information required to do updates and deletes (As for any REST Api). 

I'm fairly new to this so hope this is not too far off?

**Comment by jmdobry** [comment]
So I've been thinking about this some more, and I think it might work for this to just be an adapter. The adapter just might need to have some additional methods outside of the standard adapter methods.

**Comment by BlairAllegroTech** [comment]
We are trying to get JsData working with our Json Api backend.
So far we have not written a custom adapter but are doing everything via custom serialize and deserialize functions and the HttpAdapter. However we have come across a problem mentioned in an earlier post above. Where we have partially populated resource in the in memory data store. e.g. Just a resource with an Id (Json Api resource reference). We would like to be able to configure calls to DS#get to go to the adapter when our resource is in this state instead of returning the up populated resource.

I have tried to add an **afterInject** handler to look at the data returned in the Http response and was trying to configure the resource to always go to the adapter when the items is a Model Reference. I can see the option on DS#find to **bypassCache**. Can we configure this on the object at run time, e.g. based of the type of data that was retuned from the Http rest call.

Do you have any suggestions as to the best way to handle this?

```javascript
afterInject: (resource: any, data: any) => {

                var resourceConfig = <JSData.DSConfiguration>resource;
                if (data.constructor == Array) {
                    
                    for (var i = 0; i < data.length; ++i) {
                        var msg = "Resource: " + resource['class'] + "#" + data[i]["Id"] + ". ";

                        // Property set by our custom serializer
                        if (data[i]['IS_MODEL_REFERENCE']) {                            
                            console.log(msg + "Is Model Reference...");
                            resourceConfig.bypassCache =  true;
                            
                        } else {
                            console.log(msg);
                            resourceConfig.bypassCache = false;
                        }

                    }

                } else {                    
                    var msg = "Resource: " + resource['class'] + "#" + data["Id"] + ". ";
                    // Property set by our custom serializer
                    if (data['IS_MODEL_REFERENCE']) {
                        console.log(msg + "Is Model Reference, bypass cache...");
                        resourceConfig.bypassCache = true;
                    } else {
                        console.log(msg + "enable cache...");
                        resourceConfig.bypassCache = false;
                    }

                }                
            }
```

**Comment by jmdobry** [comment]
`bypassCache` can be set globally, per-resource, and during the method call.

**Comment by mustela** [comment]
Is there any update on this integration? Would be awesome to have it.

**Comment by BlairAllegroTech** [comment]
I have had a flu these summer holidays and so have been hard at work in this. I haven't checked it in yet. Status is, have all the current HTTPAdapter tests passing plus some more advanced JsonApi related tests also. There is still plenty more testing to be done. But i will be checking in some code to git soon. Will keep you posted

**Comment by jmdobry** [comment]
nice

**Comment by BlairAllegroTech** [comment]
I've checked some code in, I need to further document how to use it.
There are still more scenarios id like it to support of course... But am now coming up against my limited knowledge of js-data, so any help would be appreciated.

I've used the name js-data-jsonapi not sure if this is ok or not, as it is not actually a js-data project??

see : https://github.com/BlairAllegroTech/js-data-jsonapi

**Comment by jmdobry** [comment]
@BlairAllegroTech This looks fantastic! Let me know if you'd like to make `js-data-jsonapi` part of the official `js-data` GitHub organization. You would, of course, join the organization yourself and be in charge of the repository and maintain copyright. I could help with more general things like build scripts, getting code coverage and CircleCI set up, docs, publishing the package, etc. 

On another note, TypeScript is great, but when a library is authored in TypeScript, it becomes a barrier to the [85% of potential contributors who don't use TypeScript](https://ponyfoo.com/articles/javascript-developer-survey-results#which-compile-to-javascript-languages-do-you-use-if-any). Food for thought.

**Comment by jsenecal** [comment]
@BlairAllegroTech This is amazing - I was looking for something like this for ages - Do you plan on using this on a project in production ?

**Comment by BlairAllegroTech** [comment]
The concepts come from a project that we are currently working on. Not in production yet. But we had never created our own adapter and so does not quite integrate as much with js-data. When we get time. We will probably switch over to use this.

**Comment by BlairAllegroTech** [comment]
@jmdobry Sorry for the late reply....
I think i should make a bit more progress in getting this up and running and further tested before looking at the possibility of moving it into js-data.

Until then i would prefer to stay in typescript.
NOTE : The npm build include the typescript compile and lint so no special tools are required..

If i have some questions relating to the js-data API and how i am using it where do you suggest i post them.

**Comment by jmdobry** [comment]
https://js-data.slack.com/messages/js-data/

**Comment by BlairAllegroTech** [comment]
Check out the project at
https://github.com/BlairAllegroTech/js-data-jsonapi

**Comment by cottrellio** [comment]
Any movement on supporting JSON-API? We are looking at using this with AureliaJS, but it'd be helpful to have a spec to adhere to for what the endpoints should look like (like ember-data). Thanks

**Comment by GenGol** [comment]
Any updates on adding a wrapper/plugin/feature to handle JSON API 1.0

**Comment by bk-brandung** [comment]
I wonder why this isn't part of js-data. Isn't a JSON/REST API not the quasi standard for single page web application to server communication ? 😮

**Comment by jmdobry** [comment]
After 3 years of taking support requests on this project and seeing how hundreds of developers structure their APIs, I promise you, there is no standard or even a quasi standard. JSData itself chooses a default pattern that in my experience provides the greatest coverage of what Devs actually do. I believe JSData covers 95% of use cases without needing extensive configuration changes.

The term "JSON/REST API" as you used it is extremely overloaded, loosely meaning a resource-based CRUD-like pattern that uses JSON as the message format (but in my experience just means "whatever the devs decided to patch together"). JSData absolutely supports "JSON/REST API", providing the flexibility necessary for devs to map JSData to the idiosyncrasies of their APIs.

The JSON API 1.0 spec mentioned in this issue is a specific incarnation of "JSON/REST API" that not everybody uses. JSData can be configured to work with a JSON API 1.0 spec'd API, but not out-of-box. If I were to tie JSData to the JSON API 1.0 spec then I'd have to change my claim to "JSData covers 20% of use cases, and ~~to cover others requires huge configuration changes~~ doesn't really support other use cases."

**Comment by phortx** [comment]
Thank you for your answer, @jmdobry and thanks for the amazing work you did with jsdata!

You're right, everyone fiddles together what they think might be right for a JSON/REST API. I don't want to exclude them. It would just be nice if I have an standard JSON/REST API like spec'd in the spec and I can just put in some Adapter without additional work to consume that API.

That would lead to more developers creating standard APIs and it would be easier for the guys how make this stuff the "right" way.

I wont change anything in the existing Adapters. Rather I'd create a new adapter or an opt-in option for the existing http adapter to consume standard conform JSON/REST APIs without changing the de/serialization functions (like I did)

**Comment by jmdobry** [comment]
Have you looked at https://github.com/BlairAllegroTech/js-data-jsonapi?

**Comment by bk-brandung** [comment]
@jmdobry Yes, but it seems not v3 compatible
